File: elixir/bob/lib/bob.ex
defmodule Bob do

  defp containsChar?(s) do
     true in Enum.map(String.split(s,""), fn (x) -> String.match?(x,~r/^\p{L}$/u) end)
  end
  defp shouting?(s) do String.upcase(s) == s && containsChar?(s) end
  defp question?(s) do String.last(s) == "?" end
  defp silent?(s) do String.trim(s) == "" end
  def hey(input) do
    cleaned = String.trim(input)
    cond do
      silent?(cleaned) -> "Fine. Be that way!"
      shouting?(cleaned) && question?(cleaned)  -> "Calm down, I know what I'm doing!"
      shouting?(cleaned) -> "Whoa, chill out!"
      question?(cleaned) -> "Sure."
      true -> "Whatever."
    end
  end
end

File: javascript/bob/bob.js
export const hey = (message ) => {
  const isQuestion  = (x) => x.trim()[x.trim().length-1] == '?'
  const hasLetter  = (x) => [...x].map(x => (new Set("ABCDEFGHIJKLMNOPQRSTUVWXYZ").has(x))).some(x => x==true)
  const isShout = (x) => x.toUpperCase() == x && hasLetter(x)
  const isSilence = (x) => x.trim() == ""

  if (isSilence(message)) return 'Fine. Be that way!'
  if (isShout(message) && isQuestion(message)) return "Calm down, I know what I'm doing!"
  if (isShout(message)) return 'Whoa, chill out!'
  if (isQuestion(message)) return 'Sure.'
  return 'Whatever.'
};
